miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultArrayService.java

```diff
@@ -0,0 +1,238 @@
+@Transactional(rollbackFor = Exception.class)
+public class DefaultArrayService implements ArrayService {
+
+  @Autowired
+  private AuthorizationManager authorizationManager;
+
+  @Autowired
+  private ArrayStore arrayStore;
+  
+  @Autowired
+  private SampleService sampleService;
+
+  public AuthorizationManager getAuthorizationManager() {
+    return authorizationManager;
+  }
+
+  public void setAuthorizationManager(AuthorizationManager authorizationManager) {
+    this.authorizationManager = authorizationManager;
+  }
+
+  public ArrayStore getArrayStore() {
+    return arrayStore;
+  }
+
+  public void setArrayStore(ArrayStore arrayStore) {
+    this.arrayStore = arrayStore;
+  }
+
+  public SampleService getSampleService() {
+    return sampleService;
+  }
+
+  public void setSampleService(SampleService sampleService) {
+    this.sampleService = sampleService;
+  }
+
+  @Override
+  public long count(Consumer<String> errorHandler, PaginationFilter... filter) throws IOException {
+    return arrayStore.count(errorHandler, filter);
+  }
+
+  @Override
+  public List<Array> list(Consumer<String> errorHandler, int offset, int limit, boolean sortDir, String sortCol, PaginationFilter... filter)
+      throws IOException {
+    return arrayStore.list(errorHandler, offset, limit, sortDir, sortCol, filter);
+  }
+
+  @Override
+  public Array get(long arrayId) throws IOException {
+    return arrayStore.get(arrayId);
+  }
+
+  @Override
+  public long save(Array array) throws IOException {
+    loadChildEntities(array);
+    if (array.getId() == Array.UNSAVED_ID) {
+      return create(array);
+    } else {
+      return update(array);
+    }
+  }
+
+  private long create(Array array) throws IOException {
+    setChangeDetails(array);
+    validateChange(array, null);
+    return arrayStore.save(array);
+  }
+
+  private long update(Array array) throws IOException {
+    Array managed = get(array.getId());
+    validateChange(array, managed);
+    applyChanges(array, managed);
+    setChangeDetails(managed);
+    return arrayStore.save(managed);
+  }
+
+  private void loadChildEntities(Array array) throws IOException {
+    if (array.getArrayModel() != null) {
+      array.setArrayModel(getArrayModel(array.getArrayModel().getId()));
+    }
+    Map<String, Sample> samples = new HashMap<>();
+    for (Entry<String, Sample> entry : array.getSamples().entrySet()) {
+      // samples not found will end up as null mapped to a position, which will cause validation error
+      samples.put(entry.getKey(), sampleService.get(entry.getValue().getId()));
+    }
+    array.setSamples(samples);
+  }
+
+  /**
+   * Checks submitted data for validity, throwing a ValidationException containing all of the errors if invalid
+   * 
+   * @param array submitted Array to validate
+   * @param beforeChange the already-persisted Array before changes
+   * @throws IOException
+   */
+  private void validateChange(Array array, Array beforeChange) throws IOException {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 18 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)